File: app/modules/users/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class DashboardUserResponse(DashboardModel):
    id: str
    username: str
    role: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class DashboardUsersResponse(DashboardModel):
    users: list[DashboardUserResponse] = Field(default_factory=list)


class DashboardUserCreateRequest(DashboardModel):
    username: str = Field(min_length=1, max_length=128)
    password: str = Field(min_length=8, max_length=256)
    role: str = Field(pattern=r"^(admin|user)$")


class DashboardUserUpdateRequest(DashboardModel):
    username: str | None = Field(default=None, min_length=1, max_length=128)
    password: str | None = Field(default=None, min_length=8, max_length=256)
    role: str | None = Field(default=None, pattern=r"^(admin|user)$")
    is_active: bool | None = None


class DashboardUserDeleteResponse(DashboardModel):
    status: str

File: frontend/src/features/users/schemas.ts
import { z } from "zod";

export const DashboardUserRoleSchema = z.enum(["admin", "user"]);

export const DashboardUserSchema = z.object({
  id: z.string(),
  username: z.string(),
  role: DashboardUserRoleSchema,
  isActive: z.boolean(),
  createdAt: z.string().datetime({ offset: true }),
  updatedAt: z.string().datetime({ offset: true }),
});

export const DashboardUsersResponseSchema = z.object({
  users: z.array(DashboardUserSchema),
});

export const DashboardUserCreateRequestSchema = z.object({
  username: z.string().min(1).max(128),
  password: z.string().min(8).max(256),
  role: DashboardUserRoleSchema,
});

export const DashboardUserUpdateRequestSchema = z.object({
  username: z.string().min(1).max(128).optional(),
  password: z.string().min(8).max(256).optional(),
  role: DashboardUserRoleSchema.optional(),
  isActive: z.boolean().optional(),
});

export const DashboardUserDeleteResponseSchema = z.object({
  status: z.string(),
});

export type DashboardUserRole = z.infer<typeof DashboardUserRoleSchema>;
export type DashboardUser = z.infer<typeof DashboardUserSchema>;
export type DashboardUserCreateRequest = z.infer<typeof DashboardUserCreateRequestSchema>;
export type DashboardUserUpdateRequest = z.infer<typeof DashboardUserUpdateRequestSchema>;
